File: commands/stats/heatmap.py
# https://github.com/viktorexe/terristats-discord-bot
import discord
from discord.ext import commands
from discord import app_commands
from datetime import datetime, timezone
from collections import Counter
import re
from database import db

class Heatmap(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @app_commands.command(name='heatmap', description='Show clan activity patterns by hour')
    async def heatmap(self, interaction: discord.Interaction, clan: str):
        try:
            await interaction.response.defer()
            
            clan_upper = clan.upper().strip()
            
            query = {
                'clan_name': {'$regex': f'^{re.escape(clan_upper)}$', '$options': 'i'}
            }
            
            pipeline = [
                {'$match': query},
                {'$group': {
                    '_id': None,
                    'times': {'$push': '$time'},
                    'total_points': {'$sum': '$player_count'}
                }}
            ]
            
            result = await db.db.clanwins.aggregate(pipeline).to_list(1)
            if not result:
                embed = discord.Embed(
                    title=f"[{clan_upper}] Activity Heatmap",
                    description="No wins found for this clan.",
                    color=0xff0000
                )
                await interaction.followup.send(embed=embed)
                return
            
            data = result[0]
            times = data.get('times', [])
            total_points = data.get('total_points', 0)
            
            hour_counter = Counter()
            for time_str in times:
                if time_str:
                    try:
                        dt = datetime.strptime(time_str, '%a, %d %b %Y %H:%M:%S %Z')
                        hour_counter[dt.hour] += 1
                    except:
                        pass
            
            if not hour_counter:
                embed = discord.Embed(
                    title=f"[{clan_upper}] Activity Heatmap",
                    description="Unable to parse activity times.",
                    color=0xff0000
                )
                await interaction.followup.send(embed=embed)
                return
            
            max_wins = max(hour_counter.values())
            heatmap_lines = []
            
            for hour in range(24):
                wins_count = hour_counter.get(hour, 0)
                bar_length = int((wins_count / max_wins) * 10) if max_wins > 0 else 0
                bar = '█' * bar_length + '░' * (10 - bar_length)
                heatmap_lines.append(f"{hour:02d}:00 {bar} {wins_count}")
            
            col1 = '\n'.join(heatmap_lines[0:8])
            col2 = '\n'.join(heatmap_lines[8:16])
            col3 = '\n'.join(heatmap_lines[16:24])
            
            embed = discord.Embed(
                title=f"[{clan_upper}] Activity Heatmap",
                description=f"Peak activity patterns (UTC)",
                color=0x2b2d31
            )
            
            embed.add_field(name="00:00 - 07:00", value=f"```{col1}```", inline=True)
            embed.add_field(name="08:00 - 15:00", value=f"```{col2}```", inline=True)
            embed.add_field(name="16:00 - 23:00", value=f"```{col3}```", inline=True)
            
            top_hours = hour_counter.most_common(3)
            peak_text = '\n'.join([f"{hour:02d}:00 - {count} wins" for hour, count in top_hours])
            peak_text += f"\n\nTotal Points: **{total_points}**"
            embed.add_field(name="Peak Hours", value=peak_text, inline=False)
            
            total_games = await db.db.clanwins.count_documents({})
            embed.set_footer(text=f"Data from 18 Nov 2025 | {total_games:,} games tracked")
            await interaction.followup.send(embed=embed)
            
        except Exception as e:
            print(f"Error in heatmap: {e}")
            try:
                await interaction.followup.send(embed=discord.Embed(
                    title="Error", description="Failed to generate heatmap.", color=0xff0000
                ))
            except:
                pass

async def setup(bot):
    await bot.add_cog(Heatmap(bot))

File: commands/stats/heatmap.js
// https://github.com/viktorexe/terristats-discord-bot
// /heatmap - text-based activity heatmap showing a clan's wins per UTC hour.
const { SlashCommandBuilder, EmbedBuilder } = require('discord.js');
const db = require('../../database');
const { exactMatch, gamesFooter } = require('../../utils');

module.exports = {
  data: new SlashCommandBuilder()
    .setName('heatmap')
    .setDescription('Show clan activity patterns by hour')
    .addStringOption((o) =>
      o.setName('clan').setDescription('Clan name').setRequired(true),
    ),

  async execute(interaction) {
    try {
      await interaction.deferReply();

      const clan = interaction.options.getString('clan');
      const clanUpper = clan.toUpperCase().trim();
      const query = { clan_name: exactMatch(clanUpper) };

      const pipeline = [
        { $match: query },
        {
          $group: {
            _id: null,
            times: { $push: '$time' },
            total_points: { $sum: '$player_count' },
          },
        },
      ];

      const result = await db.collection('clanwins').aggregate(pipeline).toArray();
      if (!result.length) {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle(`[${clanUpper}] Activity Heatmap`)
              .setDescription('No wins found for this clan.')
              .setColor(0xff0000),
          ],
        });
        return;
      }

      const data = result[0];
      const times = data.times || [];
      const totalPoints = data.total_points || 0;

      // Count wins per hour by parsing the RFC-2822-style time string.
      const hourCounter = new Array(24).fill(0);
      for (const timeStr of times) {
        if (!timeStr) continue;
        const parsed = new Date(timeStr);
        if (!Number.isNaN(parsed.getTime())) {
          hourCounter[parsed.getUTCHours()] += 1;
        }
      }

      const maxWins = Math.max(...hourCounter);
      if (maxWins === 0) {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle(`[${clanUpper}] Activity Heatmap`)
              .setDescription('Unable to parse activity times.')
              .setColor(0xff0000),
          ],
        });
        return;
      }

      const heatmapLines = [];
      for (let hour = 0; hour < 24; hour += 1) {
        const winsCount = hourCounter[hour];
        const barLength = maxWins > 0 ? Math.floor((winsCount / maxWins) * 10) : 0;
        const bar = '█'.repeat(barLength) + '░'.repeat(10 - barLength);
        heatmapLines.push(`${String(hour).padStart(2, '0')}:00 ${bar} ${winsCount}`);
      }

      const col1 = heatmapLines.slice(0, 8).join('\n');
      const col2 = heatmapLines.slice(8, 16).join('\n');
      const col3 = heatmapLines.slice(16, 24).join('\n');

      const embed = new EmbedBuilder()
        .setTitle(`[${clanUpper}] Activity Heatmap`)
        .setDescription('Peak activity patterns (UTC)')
        .setColor(0x2b2d31)
        .addFields(
          { name: '00:00 - 07:00', value: `\`\`\`${col1}\`\`\``, inline: true },
          { name: '08:00 - 15:00', value: `\`\`\`${col2}\`\`\``, inline: true },
          { name: '16:00 - 23:00', value: `\`\`\`${col3}\`\`\``, inline: true },
        );

      // Top 3 busiest hours (only hours that actually had wins, matching the
      // original Counter.most_common behaviour).
      const topHours = hourCounter
        .map((count, hour) => [hour, count])
        .filter(([, count]) => count > 0)
        .sort((a, b) => b[1] - a[1])
        .slice(0, 3);
      let peakText = topHours
        .map(([hour, count]) => `${String(hour).padStart(2, '0')}:00 - ${count} wins`)
        .join('\n');
      peakText += `\n\nTotal Points: **${totalPoints}**`;
      embed.addFields({ name: 'Peak Hours', value: peakText, inline: false });

      const totalGames = await db.collection('clanwins').countDocuments({});
      embed.setFooter({ text: gamesFooter(totalGames) });
      await interaction.editReply({ embeds: [embed] });
    } catch (err) {
      console.log(`Error in heatmap: ${err.message}`);
      try {
        await interaction.editReply({
          embeds: [
            new EmbedBuilder()
              .setTitle('Error')
              .setDescription('Failed to generate heatmap.')
              .setColor(0xff0000),
          ],
        });
      } catch (e) {
        // ignore
      }
    }
  },
};
